quinn/src/platform/unix.rs

```diff
@@ -290,12 +290,40 @@ fn prepare_msg(
     hdr.msg_control = ctrl.0.as_mut_ptr() as _;
     hdr.msg_controllen = CMSG_LEN as _;
     let mut encoder = unsafe { cmsg::Encoder::new(hdr) };
+
     let ecn = transmit.ecn.map_or(0, |x| x as libc::c_int);
     if transmit.destination.is_ipv4() {
         encoder.push(libc::IPPROTO_IP, libc::IP_TOS, ecn as IpTosTy);
     } else {
         encoder.push(libc::IPPROTO_IPV6, libc::IPV6_TCLASS, ecn);
     }
+
+    if let Some(ip) = &transmit.dst_ip {
+        if cfg!(target_os = "linux") {
+            match ip {
+                IpAddr::V4(v4) => {
+                    let pktinfo = libc::in_pktinfo {
+                        ipi_ifindex: 0,
+                        ipi_spec_dst: libc::in_addr { s_addr: 0 },
+                        ipi_addr: unsafe {
+                            *(v4 as *const Ipv4Addr as *const () as *const libc::in_addr)
```

Comment:
Is this the correct field? `man 7 ip` says:

> If IP_PKTINFO is passed to sendmsg(2) and ipi_spec_dst is not zero, then it is used as the local  source address for the routing table lookup and for setting up IP source route options.

I'm confused about the distinction between the two address fields in general.

Reply:
I was confused too. But testing seemed to indicate it is doing the right thing. But I admit I haven't tested the IPv4 version on the host which is more susceptible to the issue. That one uses the dual-stack/IPv6 option

Reply:
I did a bit more validation here, and tested both approaches. I modified the bulk benchmark to use IPv4 only, and configured the client to manually connect to the eth1 IP of my host (not loopback). This works with setting `ipi_spec_dst`, but times out when setting `ipi_addr`. So for the sending side `ipi_spec_dst` seems like the right field, which is also indicated by the documentation.

Only weird that for reception `ipi_addr` delivers the expected value, and is rather undocumented.